### DISCUSSION TITLE
Cookies are mandatory for hosts!!

### DISCUSSION THREAD

**Discussion by sagantech** [free-audio/clap#173] [General]
So I spent a large part of my Saturday debugging a crash in my host immediately after an automation event was sent to a plugin (TAL UNO LX). 100% reproducible. It turns out that the host **MUST** propagate the "cookie" field of all parameter automations or else certain plugins will crash or worse. The "cookie" field should be better documented and is currently documented as:

>   // This value is optional and set by the plugin.
   // Its purpose is to provide a fast access to the plugin parameter:
   //
   //    Parameter *p = findParameter(param_id);
   //    param_info->cookie = p;
   //
   //    /* and later on */
   //    Parameter *p = (Parameter *)cookie;
   //
   // It is invalidated on clap_host_params->rescan(CLAP_PARAM_RESCAN_ALL) and when the plugin is
   // destroyed.

This value is **not** optional for a host.

**Comment by robbert-vdh** [comment]
Hosts should propagate these cookies in the parameter events since it's a useful optimization, but I also feel like the plugin should be able to handle missing cookies. It's still a pointer, and you should never assume a pointer is not null without checking.

**Reply by baconpaul** [reply]
I completely disagree with this @robbert-vdh 

the cookie is a pointer that is handed by the plugin so it can do internal state management and optimization. The only activity we could take if cookies were null in surge is "don't run". They are not optional.

@sagantech I agree that documentation can be better to make the importance of these much more obvious!

**Reply by sagantech** [reply]
I agree with @robbert-vdh about using a pointer without null checking. It would be better for the host not to crash (inside the plug-in) in that case even if the plug-in is non-functional. 
The main point however is as @baconpaul states: 

> documentation can be better to make the importance of these much more obvious!

**Reply by baconpaul** [reply]
So to be pedantic: we aren't using a pointer without null checking. We are using a pointer we handed to you to give back to us.

And while a null check is easy I guess a defensive check for any other random value isn't right? The job of the host is to return the pointer which was given to it from the plugin! I don't know how to check if it doesn't without maintaining a map from param id to cookie, which is what the cookie avoids.

So I don't think there is a useful check the plugin can do which is defensive other than make exactly the data structure the cookie exists to avoid using!

**Reply by robbert-vdh** [reply]
> And while a null check is easy I guess a defensive check for any other random value isn't right?

If the cookie field contains garbage values from uninitialized memory then the host triggered UB (always zero out/initialize your structs), so you can't catch that. So that's a host bug.

**Reply by sagantech** [reply]
I agree.

**Comment by sagantech** [comment]
I don't think this is an issue if the comment in the header file is changed.

**Reply by baconpaul** [reply]
Totally agree!

**Comment by baconpaul** [comment]
Here's a proposed rewrite of that comment

```
// This value is optionally set by the plugin. If, however, the plugin
// chooses to set it the host *must* set that value on every subsequent
// event which references a parameter. The host should view this value as
// a required, plugin provided, opaque, immutable value for all subsequent communications.
//
// When `clap_host_params->rescan(CLAP_PARAM_RESCAN_ALL)` the cookie is
// invalidated and the plugin can provide a new cookie value (or the same cookie value).
// After the plugin is destroyed, the cookie is no longer valid. 
//
// The host should make no assumptions about the cookie, what it points to, or if
// any value of the cookie, including nullptr, is special.
//
// The purpose to allow the plugin fast access to an internal plugin parameter
// data structure. For instance, if a plugin has its own internal class "MyParam"
// you could imagine code during the CLAP setup like
//
// auto *p = findParameter(param_id); 
// param_info->cookie = p;
//
// and then subsequently
// 
// auto *p = (MyParam *)cookie;
//
// This allows the plugin to avoid maintaining and using maps from plugin ids to parameter objects 
// during frequent events like parameter updates.
```

**Reply by sagantech** [reply]
F'n beautiful.

**Reply by baconpaul** [reply]
OK! I pushed a PR for that against the next branch. https://github.com/free-audio/clap/pull/174

Thanks!

**Reply by TaleTN** [reply]
What does "optionally set by the plugin" mean? Shouldn't the plugin always set the cookie? It could set it to NULL if it doesn't need it, but that's up to the plugin, right? Or is the host required to set the cookie to NULL before calling clap_plugin_params.get_info()?

BTW, it might also be a good idea to document what it means for the host to "invalidate" the cookie. If the host truely "should make no assumptions about the cookie", then it can't actually invalidate it either.

**Reply by sagantech** [reply]
"optionally" is a plug-in-centric view. The host MUST propagate the cookie when automating any parameters because some plug-ins assume the cookie that they set to always be there. The plug-in can set the cookie to zero but the host MUST pass the cookie to the plug-in if it does not want to crash.

**Reply by sagantech** [reply]
I am confident that "invalidate" simply means not to propagate the old cookie.

**Reply by baconpaul** [reply]
The host can’t invalidate the cookie other than by calling rescan in which case the new cookie values are in play 

the plugin may choose to not change the cookie value hence “optionally” - I think a well behaved host would initialize the cookie to nullptr before calling param info but it doesn’t have to

you can see a later draft of the comments in the pull req in this repo

**Reply by TaleTN** [reply]
> The host can’t invalidate the cookie other than by calling rescan in which case the new cookie values are in play

Thanks for explaining, that makes sense.

> the plugin may choose to not change the cookie value hence “optionally” - I think a well behaved host would initialize the cookie to nullptr before calling param info but it doesn’t have to

So, what happens on rescan, does it simply call clap_plugin_params.get_info() again? If so, then how does the plugin know that the cookie was already initialized or not?

> you can see a later draft of the comments in the pull req in this repo

Thanks, I saw it.

**Reply by baconpaul** [reply]
> > the plugin may choose to not change the cookie value hence “optionally” - I think a well behaved host would initialize the cookie to nullptr before calling param info but it doesn’t have to
> 
> So, what happens on rescan, does it simply call clap_plugin_params.get_info() again? If so, then how does the plugin know that the cookie was already initialized or not?

Basically the host always assumes the plugin is authoritative on the cookie for the entire life of the plugin except a rescan, and then yes it has to call get_info again to get the new plugin values. (That's basically what the rescan forces).

**Reply by TaleTN** [reply]
How does the plugin know that get_info() is being called the very first time, when it should initialize the cookie, or that it's being called after a rescan, when the cookie already has been initialized?

**Reply by baconpaul** [reply]
Why would it care?

**Reply by sagantech** [reply]
The cookie is usually a pointer so it will be the same every time it is initialized (unless your parameters have changed)

**Reply by baconpaul** [reply]
yeah exactly.

The mental model to have with cookies (although plugins can do anything they want) is that the plugin has an internal list of parameter objects. There's only one place in the clap API where the parameters are addressed by index and that's get_info so that's where you hand out IDs (which are what is used everywhere else to refer to parameters, including all persistence, streaming, automation).

To avoid plugins having to do ID -> object lookups using an internal map, the plugins can also hand a free pointer to the host for later reference, so the full address of a parameter is (id, cookie*) basically, although the plugin can ignore cookie.

In a rescan, the plugin can change parameter count. So for instance the CLAP API allows a plugin to tell the host to request a full param rescan in which case it may add params - those would have new cookies. Or remove them - that would invalidate old cookies. 

But the by far most common case with CLAP will be that the cookie is a pointer to the juce::parameter or iplug::parameter object or what not,.

**Reply by TaleTN** [reply]
I feel I'm going in circles, but let's have at least one more try: :-)

So *setting* the cookie in get_info() isn't optional, but *using* the cookie elsewhere in the plugin is optional. Correct?

**Reply by sagantech** [reply]
If you are developing a plug-in (not a host) setting the cookie IS optional.

**Reply by TaleTN** [reply]
But isn't the host still required to propagate the cookie, even if the plugin has chosen to not set/initialize it? If so, then this will probably work just fine in practice, but strictly speaking, AFAIK reading an uninitialized variable is undefined behavior in C.

**Reply by baconpaul** [reply]
The host may want to initialize the info structure before it calls the get info
But a void* is just a value. The host gets it and sends it back.
